File: packages/js/graph.js
class Graph {
    constructor() {
        this.adjacencyList = new Map();
    }
    addVertex(v) {
        this.adjacencyList.set(v, []);
    }
    addEdge(v, w, weight = 0) {
        if (!this.adjacencyList.has(v)) this.addVertex(v);
        if (!this.adjacencyList.has(w)) this.addVertex(w);
        this.adjacencyList.get(v).push({ node: w, weight });
        // Undirected graph
        this.adjacencyList.get(w).push({ node: v, weight });
    }
    bfs(startingNode) {
        let visited = new Set();
        let queue = [startingNode];
        let result = [];
        visited.add(startingNode);

        while (queue.length > 0) {
            let current = queue.shift();
            result.push(current);
            
            if(this.adjacencyList.has(current)) {
                let neighbors = this.adjacencyList.get(current);
                for (let neighbor of neighbors) {
                    if (!visited.has(neighbor.node)) {
                        visited.add(neighbor.node);
                        queue.push(neighbor.node);
                    }
                }
            }
        }
        return result;
    }
    dfs(startingNode) {
        let visited = new Set();
        let result = [];
        const dfsHelper = (vertex) => {
            visited.add(vertex);
            result.push(vertex);
            if(this.adjacencyList.has(vertex)) {
                let neighbors = this.adjacencyList.get(vertex);
                for (let neighbor of neighbors) {
                    if (!visited.has(neighbor.node)) {
                        dfsHelper(neighbor.node);
                    }
                }
            }
        };
        dfsHelper(startingNode);
        return result;
    }
    kruskalMST() {
        let edges = [];
        this.adjacencyList.forEach((neighbors, u) => {
            for (let v of neighbors) {
                if (u < v.node) {
                    edges.push({ u, v: v.node, weight: v.weight });
                }
            }
        });
        edges.sort((a, b) => a.weight - b.weight);

        let parent = {};
        const find = (i) => {
            if (parent[i] === i) return i;
            return find(parent[i]);
        };
        const union = (i, j) => {
            let rootI = find(i);
            let rootJ = find(j);
            parent[rootI] = rootJ;
        };

        this.adjacencyList.forEach((_, v) => parent[v] = v);

        let mst = [];
        for (let edge of edges) {
            let rootU = find(edge.u);
            let rootV = find(edge.v);
            if (rootU !== rootV) {
                mst.push(edge);
                union(rootU, rootV);
            }
        }
        return mst;
    }
}

module.exports = { Graph };

File: packages/python/graph.py
class Graph:
    def __init__(self):
        self.adj = {}
    
    def add_edge(self, u, v, weight=1):
        if u not in self.adj: self.adj[u] = []
        if v not in self.adj: self.adj[v] = []
        self.adj[u].append((v, weight))
        self.adj[v].append((u, weight)) # Undirected
        
    def bfs(self, start):
        if start not in self.adj: return []
        visited = set([start])
        queue = [start]
        res = []
        while queue:
            node = queue.pop(0)
            res.append(node)
            for neighbor, _ in self.adj.get(node, []):
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        return res

    def dfs(self, start):
        if start not in self.adj: return []
        visited = set()
        res = []
        def _dfs(node):
            visited.add(node)
            res.append(node)
            for neighbor, _ in self.adj.get(node, []):
                if neighbor not in visited:
                    _dfs(neighbor)
        _dfs(start)
        return res

    def kruskal_mst(self):
        edges = []
        for u in self.adj:
            for v, w in self.adj[u]:
                if u < v:
                    edges.append((w, u, v))
        edges.sort()
        
        parent = {}
        def find(i):
            if parent[i] == i:
                return i
            parent[i] = find(parent[i])
            return parent[i]
            
        def union(i, j):
            root_i = find(i)
            root_j = find(j)
            if root_i != root_j:
                parent[root_i] = root_j
                
        for node in self.adj:
            parent[node] = node
            
        mst = []
        for w, u, v in edges:
            if find(u) != find(v):
                union(u, v)
                mst.append((u, v, w))
        return mst
